Source/WebCore/dom/ElementTextDirection.cpp

```diff
@@ -0,0 +1,325 @@
+    if (equalLettersIgnoringASCIICase(value, "rtl"_s))
+        return TextDirectionality::RTL;
+    if (equalLettersIgnoringASCIICase(value, "auto"_s))
+        return TextDirectionality::Auto;
+    return TextDirectionality::Undefined;
+}
+
+static TextDirectionality elementTextDirectionality(const Element& element)
+{
+    return parseTextDirectionality(element.attributeWithoutSynchronization(dirAttr));
+}
+
+bool elementHasValidTextDirectionality(const Element& element)
+{
+    return elementTextDirectionality(element) != TextDirectionality::Undefined;
+}
+
+static bool elementHasOwnValidTextDirectionality(const Element& element)
+{
+    return is<HTMLBDIElement>(element) || elementHasValidTextDirectionality(element);
+}
+
+bool elementHasOwnValidAutoTextDirectionality(const Element& element)
+{
+    auto textDirectionality = elementTextDirectionality(element);
+    return textDirectionality == TextDirectionality::Auto || (is<HTMLBDIElement>(element) && textDirectionality == TextDirectionality::Undefined);
+}
+
+static void setNodeHasDirAutoFlagRecursively(Node& firstNode, bool flag)
+{
+    firstNode.setSelfOrPrecedingNodesAffectDirAuto(flag);
+
+    for (RefPtr node = firstNode.firstChild(); node; ) {
+        auto* element = dynamicDowncast<Element>(*node);
+
+        if (element && elementHasOwnValidTextDirectionality(*element)) {
+            node = NodeTraversal::nextSkippingChildren(*node, &firstNode);
+            continue;
+        }
+
+        node->setSelfOrPrecedingNodesAffectDirAuto(flag);
+        node = NodeTraversal::next(*node, &firstNode);
+    }
+}
+
+static void setElementHasDirAutoFlag(Element& element, TextDirectionality textDirectionality)
+{
+    RefPtr parent = element.parentOrShadowHostElement();
+    element.protectedDocument()->setIsDirAttributeDirty();
+
+    switch (textDirectionality) {
+    case TextDirectionality::LTR:
+    case TextDirectionality::RTL:
+        if (element.selfOrPrecedingNodesAffectDirAuto() || (parent && parent->selfOrPrecedingNodesAffectDirAuto()))
+            setNodeHasDirAutoFlagRecursively(element, false);
+        break;
+
+    case TextDirectionality::Auto:
+        setNodeHasDirAutoFlagRecursively(element, true);
+        break;
+
+    case TextDirectionality::Undefined:
+        if (element.selfOrPrecedingNodesAffectDirAuto() && !(parent && parent->selfOrPrecedingNodesAffectDirAuto()) && !is<HTMLBDIElement>(element))
+            setNodeHasDirAutoFlagRecursively(element, false);
+        break;
+    }
+}
+
+static std::optional<TextDirection> computeTextTextDirection(const String& text)
+{
+    if (auto direction = text.defaultWritingDirection())
+        return *direction == U_LEFT_TO_RIGHT ? TextDirection::LTR : TextDirection::RTL;
+    return std::nullopt;
+}
+
+static std::optional<TextDirection> computeElementDescendantsAutoTextDirection(const Element& element)
+{
+    for (RefPtr child = element.firstChild(); child; ) {
+        // Specs: Skip bdi, script, style nodes.
```

Comment:
I don't think these comments are useful.

Reply:
The algorithm is really complex. And I wanted to explicitly differentiate between the parts of the computation which I borrowed from the specs and the parts which I wrote to control the flow of the code. 

Initially I wrote this case in `computeElementTextDirection()` like this:
```
    switch (textDirectionality) {
        ...
    case TextDirectionality::Auto:
        // Specs: Return the auto directionality of element.
        return computeElementAutoTextDirection(element);
        ...
    }
    return std::nullopt;
```
But it turned out to be wrong and it is not aligned with the specs and I had to change it to be like this

```
    switch (textDirectionality) {
        ...
    case TextDirectionality::Auto:
        // Specs: Return the auto directionality of element.
        // Specs: If directionality is null, then return 'ltr'.
        return computeElementAutoTextDirection(element).value_or(TextDirection::LTR);
        ...
    }

    return std::nullopt;
```
This change fixed many layout test failures and I wanted to document this change.

Reply:
Yeah but this comment just repeats the code. It literally just skips bid, script, & style elements.

Reply:
The point is not to repeat what the code is doing. The point to document that this code was brought from the specs as is. So if the code changes accidentally, just by reading the code and the comment, we will know that it is not aligned with the specs anymore.